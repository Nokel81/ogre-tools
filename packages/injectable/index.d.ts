/// <reference types="jest" />
export interface DiContainer extends DiContainerForInjection {
  purge: (injectableKey: Injectable<any, any, any>) => void;

  permitSideEffects: (injectableKey: Injectable<any, any, any>) => void;

  override<
    InjectionInstance extends InjectionTokenInstance,
    InjectionTokenInstance,
    InstantiationParam,
  >(
    injectable: Injectable<
      InjectionInstance,
      InjectionTokenInstance,
      InstantiationParam
    >,
    instantiateStub: Instantiate<InjectionInstance, InstantiationParam>,
  ): void;

  unoverride(injectable: Injectable<any, any, any>): void;

  register(...injectables: Injectable<any, any, any>[]): void;

  deregister(...injectables: Injectable<any, any, any>[]): void;
  preventSideEffects: () => void;
}

export type Instantiate<InjectionInstance, InstantiationParam> = {
  (
    di: DiContainerForInjection,
    param: InstantiationParam extends void ? void : InstantiationParam,
  ): InjectionInstance;
};

export interface InjectionToken<InjectionInstance, InstantiationParam> {
  template: InjectionInstance;
  instantiationParameter: InstantiationParam;
  key: Symbol;
}

export interface Injectable<
  InjectionInstance extends InjectionTokenInstance,
  InjectionTokenInstance,
  InstantiationParam,
> {
  id: string;
  causesSideEffects?: boolean;
  injectionToken?: InjectionToken<InjectionTokenInstance, InstantiationParam>;
  instantiate: Instantiate<InjectionInstance, InstantiationParam>;
  lifecycle: ILifecycle<InstantiationParam>;
}

type InjectableLifecycle<InstantiationParam> = InstantiationParam extends void
  ? {
      lifecycle?: ILifecycle<void>;
    }
  : {
      lifecycle: ILifecycle<InstantiationParam>;
    };

type GetInjectableOptions<
  InjectionInstance extends InjectionTokenInstance,
  InjectionTokenInstance,
  InstantiationParam,
> = InjectableLifecycle<InstantiationParam> &
  Omit<
    Injectable<InjectionInstance, InjectionTokenInstance, InstantiationParam>,
    'lifecycle'
  >;

export function getInjectable<
  InjectionInstance extends InjectionTokenInstance,
  InjectionTokenInstance,
  InstantiationParam = void,
>(
  options: GetInjectableOptions<
    InjectionInstance,
    InjectionTokenInstance,
    InstantiationParam
  >,
): Injectable<InjectionInstance, InjectionTokenInstance, InstantiationParam>;

export function getInjectionToken<
  InjectionInstance,
  InstantiationParam = void,
>(options: {
  id: string;
}): InjectionToken<InjectionInstance, InstantiationParam>;

interface Inject {
  <InjectionInstance>(
    key:
      | Injectable<InjectionInstance, unknown, void>
      | InjectionToken<InjectionInstance, void>,
  ): InjectionInstance;

  <InjectionInstance, InstantiationParam>(
    key:
      | Injectable<InjectionInstance, unknown, InstantiationParam>
      | InjectionToken<InjectionInstance, InstantiationParam>,
    param: InstantiationParam,
  ): InjectionInstance;
}

interface InjectMany {
  <InjectionInstance>(
    key:
      | Injectable<InjectionInstance, unknown, void>
      | InjectionToken<InjectionInstance, void>,
  ): InjectionInstance[];

  <InjectionInstance, InstantiationParam>(
    key:
      | Injectable<InjectionInstance, unknown, InstantiationParam>
      | InjectionToken<InjectionInstance, InstantiationParam>,
    param: InstantiationParam,
  ): InjectionInstance[];
}

export interface DiContainerForInjection {
  inject: Inject;
  injectMany: InjectMany;
  register(...injectables: Injectable<any, any, any>[]): void;
  deregister(...injectables: Injectable<any, any, any>[]): void;
}

export interface ILifecycle<InstantiationParam> {
  getInstanceKey: (di: DiContainer, params: InstantiationParam) => any;
}

declare const storedInstanceKey: unique symbol;
declare const nonStoredInstanceKey: unique symbol;

export const lifecycleEnum: {
  singleton: {
    getInstanceKey: (di: DiContainer) => typeof storedInstanceKey;
  };

  keyedSingleton<InstantiationParam>(
    options: ILifecycle<InstantiationParam>,
  ): typeof options;

  transient: {
    getInstanceKey: (di: DiContainer) => typeof nonStoredInstanceKey;
  };
};

export function createContainer(containerId: string): DiContainer;
