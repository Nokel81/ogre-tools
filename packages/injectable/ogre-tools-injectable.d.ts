/// <reference types="jest" />
declare module '@ogre-tools/injectable' {
  type InferFromInjectable<T> = T extends Injectable<
    unknown,
    infer TInstance,
    infer TInstantiationParameter
  >
    ? [TInstance, TInstantiationParameter]
    : never;

  type TentativeTuple<T> = T extends object ? [T] : [undefined?];

  export interface DependencyInjectionContainer {
    inject<TInjectable extends Injectable<unknown, unknown, unknown>>(
      injectableKey: TInjectable,
      ...instantiationParameter: TentativeTuple<
        InferFromInjectable<TInjectable>[1]
      >
    ): InferFromInjectable<TInjectable>[0];

    inject<TInjectionToken extends InjectionToken<unknown, unknown>>(
      injectionToken: TInjectionToken,
      ...instantiationParameter: TentativeTuple<
        TInjectionToken['instantiationParameter']
      >
    ): TInjectionToken['template'];

    purge: (injectableKey: Injectable<any, any, any>) => void;

    runSetups: () => Promise<void>;

    override<TInjectable extends Injectable<unknown, unknown, unknown>>(
      injectable: TInjectable,
      instantiateStub: (
        di: DependencyInjectionContainer,
        ...instantiationParameter: TentativeTuple<
          InferFromInjectable<TInjectable>[1]
        >
      ) => InferFromInjectable<TInjectable>[0],
    ): void;
  }

  export interface ConfigurableDependencyInjectionContainer
    extends DependencyInjectionContainer {
    register(injectable: Injectable<any, any, any>): void;
    preventSideEffects: () => void;
  }

  type InjectionToken<TInstance, TInstantiationParameter> = {
    template: TInstance;
    instantiationParameter: TInstantiationParameter;
    key: Symbol;
  };

  export function getInjectionToken<
    TInstance,
    TInstantiationParameter = void,
  >(): InjectionToken<TInstance, TInstantiationParameter>;

  export interface Injectable<
    TInjectionToken,
    TInstance,
    TInstantiationParameter,
  > {
    id?: string;
    setup?: (di: DependencyInjectionContainer) => void | Promise<void>;
    causesSideEffects?: boolean;
    lifecycle: lifecycleEnum;

    instantiate: (
      di: DependencyInjectionContainer,
      instantiationParameter: TInstantiationParameter,
    ) => TInstance;

    injectionToken?: TInjectionToken;
  }

  type InferFromToken<T> = T extends InjectionToken<
    infer TInstance,
    infer TInstantiationParameter
  >
    ? [TInstance, TInstantiationParameter]
    : never;

  export function getInjectable<
    TInjectionToken extends InjectionToken<TInstance, TInstantiationParameter>,
    TInstance = TInjectionToken extends InjectionToken<any, any>
      ? InferFromToken<TInjectionToken>[0]
      : unknown,
    TInstantiationParameter = TInjectionToken extends InjectionToken<any, any>
      ? InferFromToken<TInjectionToken>[1]
      : unknown,
  >(
    options: Injectable<TInjectionToken, TInstance, TInstantiationParameter>,
  ): Injectable<TInjectionToken, TInstance, TInstantiationParameter>;

  export enum lifecycleEnum {
    singleton,
    transient,
  }

  export function createContainer(
    ...getRequireContexts: any[]
  ): ConfigurableDependencyInjectionContainer;
}
