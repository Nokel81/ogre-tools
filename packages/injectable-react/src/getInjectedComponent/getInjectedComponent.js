import React from 'react';
import constant from 'lodash/fp/constant';
import { Observer } from 'mobx-react';
import { action, observable } from 'mobx';

import { isPromise } from '@ogre-tools/fp';
import { pipeline } from '../../../fp/src/index';
import toPairs from 'lodash/fp/toPairs';
import map from 'lodash/fp/map';
import fromPairs from 'lodash/fp/fromPairs';

const { Provider: DiContextProvider, Consumer: DiContextConsumer } =
  React.createContext();

export { DiContextProvider };

export default (
    Component,
    { getPlaceholder = constant(null), getDependencies },
  ) =>
  props =>
    (
      <DiContextConsumer>
        {({ di }) => {
          const maybeAsyncDependencies = pipeline(
            di,
            getDependencies,
            synchronize,
          );

          if (!isPromise(maybeAsyncDependencies)) {
            return <Component {...maybeAsyncDependencies} {...props} />;
          }

          const observableDependencyPromise = getObservablePromise(
            maybeAsyncDependencies,
          );

          return (
            <Observer>
              {() => {
                const dependencies = observableDependencyPromise.value;

                if (!dependencies) {
                  return getPlaceholder();
                }

                return <Component {...dependencies} {...props} />;
              }}
            </Observer>
          );
        }}
      </DiContextConsumer>
    );

const getObservablePromise = asyncValue => {
  const observableObject = observable({ value: null }, null, {
    deep: false,
  });

  asyncValue.then(
    action(syncValue => {
      observableObject.value = syncValue;
    }),
  );

  return observableObject;
};

const synchronize = maybeAsyncDependencies =>
  pipeline(
    maybeAsyncDependencies,
    toPairs,

    map(([key, maybeAsyncDependency]) =>
      isPromise(maybeAsyncDependency)
        ? maybeAsyncDependency.then(syncDependency => [key, syncDependency])
        : [key, maybeAsyncDependency],
    ),

    fromPairs,
  );
