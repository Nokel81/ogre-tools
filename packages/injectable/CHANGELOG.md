# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [4.1.0](https://github.com/ogre-works/ogre-tools/compare/v4.0.0...v4.1.0) (2022-02-08)


### Features

* Expose keyed singleton as lifecycle in types ([282762e](https://github.com/ogre-works/ogre-tools/commit/282762e3b8e7998fd26474198c10592c67c89ddb))
* introduce keyedSingleton as lifecycle ([174472c](https://github.com/ogre-works/ogre-tools/commit/174472c296c8bdaf2b5ca38469d1f9b3a1963277))
* Permit primitive instantiation parameters ([b10dd82](https://github.com/ogre-works/ogre-tools/commit/b10dd82d7674952129ee5eeb31f50e0c5cb06e6c))



## [4.0.0](https://github.com/ogre-works/ogre-tools/compare/v3.2.1...v4.0.0) (2022-02-07)


### âš  BREAKING CHANGES

* force breaking change to cause major version bump in Lerna

### Bug Fixes

* fix build by using webpack made available to a package by lerna ([5b08e24](https://github.com/ogre-works/ogre-tools/commit/5b08e2472fe06514901546e8a5eb8d8664282a0c))
* force breaking change to cause major version bump in Lerna ([f0c1def](https://github.com/ogre-works/ogre-tools/commit/f0c1defeb85166ccf1907e5d6e9a36cb1cb2314a))


### Reverts

* Revert "Revert "Enhance error for injecting non-registered injectable"" ([8e24a0c](https://github.com/ogre-works/ogre-tools/commit/8e24a0c058758a4c8c50209e9702fd7a2bd792f9))
* Revert "Revert "Implement di.injectMany() as way to inject multiple registered injectables for an injection token"" ([f5d0e01](https://github.com/ogre-works/ogre-tools/commit/f5d0e013d7e079deb5fe1eba5396595850930901))
* Revert "Revert "Permit injection using only injectable or injection token. Replace "id" with "module" for better error messages."" ([fc67c38](https://github.com/ogre-works/ogre-tools/commit/fc67c38aaf4fd4bf81edbcd8dd8bc843864ba0ce))
* Revert "Revert "Remove concept of "injectable viability" for being YAGNI"" ([e4979c6](https://github.com/ogre-works/ogre-tools/commit/e4979c6ef6a002e0f8ae1da248323a2cba672049))
* Revert "Revert "Prevent public access to injection context"" ([4c88b4a](https://github.com/ogre-works/ogre-tools/commit/4c88b4a1c7db1135a2a47c5f97dd976c085bdb42))
