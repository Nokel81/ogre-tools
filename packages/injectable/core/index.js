import getInjectionToken from './src/getInjectionToken/getInjectionToken';

import getInjectable from './src/getInjectable/getInjectable';
import lifecycleEnum from './src/dependency-injection-container/lifecycleEnum';

import createContainer from './src/dependency-injection-container/createContainer';

import isInjectable from './src/getInjectable/isInjectable';
import isInjectionToken from './src/getInjectionToken/isInjectionToken';
import {
  deregistrationCallbackToken,
  injectionDecoratorToken,
  instantiationDecoratorToken,
  registrationCallbackToken,
} from './src/dependency-injection-container/tokens';

export {
  createContainer,
  getInjectable,
  getInjectionToken,
  injectionDecoratorToken,
  instantiationDecoratorToken,
  registrationCallbackToken,
  deregistrationCallbackToken,
  isInjectable,
  isInjectionToken,
  lifecycleEnum,
};
