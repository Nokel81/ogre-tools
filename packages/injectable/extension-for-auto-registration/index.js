import fs from 'fs';
import path from 'path';

import autoRegisterFor from './src/autoRegister';

export const autoRegister = autoRegisterFor({ fs, path });
