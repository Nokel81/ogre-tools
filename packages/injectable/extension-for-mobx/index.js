import {
  computedInjectManyInjectable,
  computedInjectManyWithMetaInjectable,
  registerMobX,
} from './src/computedInjectMany';

export {
  registerMobX,
  computedInjectManyInjectable,
  computedInjectManyWithMetaInjectable,
};
