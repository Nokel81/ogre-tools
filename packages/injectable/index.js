import getInjectionToken from './src/getInjectionToken/getInjectionToken';
import getInjectable from './src/getInjectable/getInjectable';
import lifecycleEnum from './src/dependency-injection-container/lifecycleEnum';
import createContainer, {
  errorMonitorInjectionToken,
} from './src/dependency-injection-container/createContainer';

export {
  getInjectionToken,
  getInjectable,
  lifecycleEnum,
  createContainer,
  errorMonitorInjectionToken,
};
