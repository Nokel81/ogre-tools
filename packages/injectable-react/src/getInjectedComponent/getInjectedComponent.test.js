import React from 'react';
import enzyme from 'enzyme';
import { setImmediate as flushMicroTasks } from 'timers';
import { createContainer, lifecycleEnum } from '@ogre-tools/injectable';
import getInjectedComponent from './getInjectedComponent';
import { DiContextProvider } from '../index';
import asyncFn from '@async-fn/jest';

const flushPromises = () => new Promise(flushMicroTasks);

const enzymeUpdate = async component => {
  component.setProps();

  await flushPromises();

  component.find('Observer').forEach(Observer => Observer.update());
};

const mountFor =
  di =>
  (node, ...rest) =>
    enzyme.mount(<DiContextProvider value={{ di }}>{node}</DiContextProvider>, {
      ...rest,
    });

describe('getInjectedComponent', () => {
  let di;
  let mount;

  beforeEach(() => {
    di = createContainer();

    mount = mountFor(di);
  });

  it('given injected component, when rendered, renders with dependencies', () => {
    di.register({
      id: 'some-injectable-id',

      lifecycle: lifecycleEnum.transient,

      getDependencies: () => ({
        someDependency: 'some-dependency-value',
      }),

      instantiate: (dependencies, props) => (
        <DumbTestComponent {...dependencies} {...props} />
      ),
    });

    const DumbTestComponent = ({ someDependency, ...props }) => (
      <div {...props}>Some content: "{someDependency}"</div>
    );

    const SmartTestComponent = getInjectedComponent('some-injectable-id');

    const component = mount(<SmartTestComponent data-some-prop-test />);

    expect(component).toMatchHtmlSnapshot();
  });

  describe('given async dependency, a placeholder and injected component, when rendered', () => {
    let component;
    let asyncDependencyMock;

    beforeEach(async () => {
      asyncDependencyMock = asyncFn();

      di.register({
        id: 'some-injectable-id',

        lifecycle: lifecycleEnum.transient,

        getDependencies: () => ({
          someDependency: asyncDependencyMock(),
        }),

        instantiate: (dependencies, props) => (
          <DumbTestComponent {...dependencies} {...props} />
        ),
      });

      const DumbTestComponent = ({ someDependency, ...props }) => (
        <div {...props}>Some content: "{someDependency}"</div>
      );

      const SmartTestComponent = getInjectedComponent('some-injectable-id', {
        getPlaceholder: () => <div data-placeholder-test />,
      });

      component = mount(<SmartTestComponent data-some-prop-test />);
    });

    it('renders', () => {
      expect(component).toMatchHtmlSnapshot();
    });

    it('has placeholder', () => {
      expect(component.find('[data-placeholder-test]')).toExist();
    });

    describe('when the dependency resolves', () => {
      beforeEach(async () => {
        await asyncDependencyMock.resolve('some-async-value');
        await enzymeUpdate(component);
      });

      it('renders', () => {
        expect(component).toMatchHtmlSnapshot();
      });

      it('has no placeholder', () => {
        expect(component.find('[data-placeholder-test]')).not.toExist();
      });
    });
  });
});
