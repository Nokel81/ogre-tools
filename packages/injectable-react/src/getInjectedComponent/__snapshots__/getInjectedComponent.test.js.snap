// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getInjectedComponent given component and sync dependencies, when rendered, renders with dependencies 1`] = `
<div
  data-some-prop-test="true"
>
  Some content: "some-injectable-value"
</div>
`;

exports[`getInjectedComponent given component, no placeholder and async dependencies, when rendered when the dependency resolves renders 1`] = `
<div
  data-some-prop-test="true"
>
  Some content: "some-async-value"
</div>
`;

exports[`getInjectedComponent given component, placeholder and async dependencies, when rendered renders as placeholder 1`] = `
<div
  data-placeholder-test="true"
/>
`;

exports[`getInjectedComponent given component, placeholder and async dependencies, when rendered when the dependency resolves renders 1`] = `
<div
  data-some-prop-test="true"
>
  Some content: "some-async-value"
</div>
`;
